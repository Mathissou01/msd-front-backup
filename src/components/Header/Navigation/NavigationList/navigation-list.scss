@use "src/styles/01-settings/settings.colors" as colors;

.c-NavigationList {
  $self: &;

  display: flex;
  flex-direction: column;
  margin: 0;
  padding: 0;
  list-style: none;

  &_desktop {
    flex-grow: 1;
    align-items: flex-start;
  }

  &__Item {
    display: flex;
    flex-direction: column;
    width: 100%;

    &:hover {
      background-color: colors.$brand-primary-light;
    }

    #{$self}_mobile & {
      height: 66px;
    }

    #{$self}_desktop & {
      flex-grow: 1;
      gap: 6px;
      align-items: center; // TODO: remove after replacing all items with NavigationListButton
      justify-content: center;
      min-height: 70px;
      max-height: 102px;
      border-bottom: 1px solid colors.$gray-200;
    }

    &:first-child:not(&_active) {
      border-top: 1px solid colors.$gray-200;
    }

    &_active {
      #{$self}_desktop & {
        background-color: colors.$brand-primary;
      }
    }
  }
}
