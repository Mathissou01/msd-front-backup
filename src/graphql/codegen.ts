import type { CodegenConfig } from "@graphql-codegen/cli";

import { config as configDotenv } from "dotenv";

configDotenv();

const config: CodegenConfig = {
  overwrite: true,
  schema: `${process.env.NEXT_PUBLIC_API_URL}/graphql`,
  documents: "./src/graphql/**/*.graphql",
  ignoreNoDocuments: true,
  generates: {
    "./src/graphql/codegen/schema.graphql": {
      plugins: ["schema-ast"],
    },
    "./src/graphql/codegen/generated-types.ts": {
      plugins: [
        {
          add: {
            content: "/* eslint-disable */",
          },
        },
        "typescript",
        "typescript-operations",
        "typescript-react-apollo",
      ],
    },
  },
  hooks: {
    afterOneFileWrite: ["prettier --write"],
  },
};

export default config;
