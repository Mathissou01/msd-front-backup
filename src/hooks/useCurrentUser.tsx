import React, {
  Dispatch,
  SetStateAction,
  createContext,
  useContext,
  useEffect,
  useState,
} from "react";
import { User } from "../lib/user";

interface UserContextValues {
  currentUser: User | null;
  setCurrentUser: Dispatch<SetStateAction<User | null>>;
  login: () => void;
  logout: () => void;
}

export const useCurrentUser = (): UserContextValues => {
  const context = useContext(UserContext);
  if (!context) {
    throw new Error();
  }
  return context;
};

const UserContext = createContext<UserContextValues>({} as UserContextValues);

export const UserProvider = ({ children }: { children: React.ReactNode }) => {
  const [currentUser, setCurrentUser] = useState<User | null>(null);

  const getUser = async () => {
    try {
      const storedUser = localStorage.getItem("user");
      if (storedUser && storedUser !== "null") {
        setCurrentUser(JSON.parse(storedUser));
      }
    } catch (error) {
      console.error(error);
    }
  };

  const login = () => {
    const user = {
      id: "1",
      email: "john.doe@gmail.com",
      firstname: "John",
      lastname: "Doe",
      activeCounter: true,
      householdSize: 1,
      userType: "particular",
      dwellingType: "house",
      addressLabel: "1 rue de la paix, 75000 Paris",
    };
    setCurrentUser(user);
    localStorage.setItem("user", JSON.stringify(user));
  };

  const logout = () => {
    setCurrentUser(null);
    localStorage.removeItem("user");
  };

  useEffect(() => {
    getUser();
  }, []);

  useEffect(() => {
    localStorage.setItem("user", JSON.stringify(currentUser));
  }, [currentUser]);

  const contextValue: UserContextValues = {
    currentUser,
    setCurrentUser,
    login,
    logout,
  };

  return (
    <UserContext.Provider value={contextValue}>{children}</UserContext.Provider>
  );
};

export default UserContext;
