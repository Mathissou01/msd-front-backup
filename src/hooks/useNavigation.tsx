import React, { useContext } from "react";

export enum ENavigationPages {
  "HomeLink" = "/",
  // Editorial
  "ComponentLinksNews" = "/actualites-evenements",
  "ComponentLinksQuizzes" = "/quiz",
  "ComponentLinksTips" = "/astuces",
  "ComponentLinksFrees" = "/contenu-libre", // Path is concatenated with name as normalized string
  "ComponentLinksContactUs" = "/contact",
  "ComponentLinksMyWasteCounter" = "/mon-compteur-dechets",
  // Services
  "ComponentLinksRecyclingGuide" = "/service/guide-tri",
  "ComponentLinksDropOffMap" = "/service/carte",
  "ComponentLinksPickUpDay" = "/service/collecte-adresse",
  "ComponentLinksRequest" = "/service/demandes",
  // Misc
  "ComponentLinksExternal" = "/", // TODO: handle external links
}

export interface INavigationContext {
  currentPage: ENavigationPages | string;
  setCurrentPage: (pageName: ENavigationPages | string) => void;
}

export const NavigationContext = React.createContext<INavigationContext>({
  currentPage: ENavigationPages.HomeLink,
  setCurrentPage: () => null,
});

export const useNavigation = () => useContext(NavigationContext);
