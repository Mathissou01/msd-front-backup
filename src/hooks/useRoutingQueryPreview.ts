import { useRouter } from "next/router";
import { TPreviewTypes } from "../lib/preview";

export function useRoutingQueryPreview<T extends string | TPreviewTypes>(
  queryParam: string,
  validateQueryParam: (param: string) => boolean,
): T | undefined {
  const router = useRouter();
  const query = router.query[queryParam];
  if (query?.toString()) {
    if (!validateQueryParam(query.toString())) {
      void router.push("/404");
      return undefined;
    }
    return query?.toString() as T;
  } else if (typeof window === undefined) {
    void router.push("/404");
  }

  return undefined;
}
