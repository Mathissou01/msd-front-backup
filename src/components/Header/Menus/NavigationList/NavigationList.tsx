import NavigationListButton from "./NavigationListButton/NavigationListButton";
import "./navigation-list.scss";

interface INavigationListProps {
  isDesktopMode: boolean;
}

export default function NavigationList({
  isDesktopMode,
}: INavigationListProps) {
  const temporaryLabels = {
    // TODO: temporary static values, remove later
    homepage: "Accueil",
    map: "Carte",
    address: "Collecte à mon adresse",
    recyclingGuide: "Guide du tri",
    demand: "Faire une demande",
    wastes: "Réduire mes déchets",
    leaf: "Valorisation des déchets",
  };

  return (
    <ul
      className={`c-NavigationList ${
        isDesktopMode ? "c-NavigationList_desktop" : "c-NavigationList_mobile"
      }`}
    >
      <li className="c-NavigationList__Item c-NavigationList__Item_active">
        <NavigationListButton
          label={temporaryLabels.homepage}
          picto="homepage"
          isDesktopMode={isDesktopMode}
          isActive={true}
        />
      </li>
      <li className="c-NavigationList__Item">
        <NavigationListButton
          label={temporaryLabels.map}
          picto="map"
          isDesktopMode={isDesktopMode}
        />
      </li>
      <li className="c-NavigationList__Item">
        <NavigationListButton
          label={temporaryLabels.address}
          picto="address"
          isDesktopMode={isDesktopMode}
        />
      </li>
      <li className="c-NavigationList__Item">
        <NavigationListButton
          label={temporaryLabels.recyclingGuide}
          picto="recyclingGuide"
          isDesktopMode={isDesktopMode}
        />
      </li>
      <li className="c-NavigationList__Item">
        <NavigationListButton
          label={temporaryLabels.demand}
          picto="demand"
          isDesktopMode={isDesktopMode}
        />
      </li>
      <li className="c-NavigationList__Item">
        <NavigationListButton
          label={temporaryLabels.wastes}
          picto="wastes"
          isDesktopMode={isDesktopMode}
        />
      </li>
      <li className="c-NavigationList__Item">
        <NavigationListButton
          label={temporaryLabels.leaf}
          picto="leaf"
          isDesktopMode={isDesktopMode}
        />
      </li>
    </ul>
  );
}
