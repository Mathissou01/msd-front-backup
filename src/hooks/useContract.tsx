import React, { useContext } from "react";
import { ContractEntity } from "../graphql/codegen/generated-types";

export interface IContractContext {
  contract: ContractEntity;
  contractId: `${number}`;
}

export const ContractContext = React.createContext<IContractContext>({
  contract: {},
  contractId: "0",
});

export const useContract = () => useContext(ContractContext);
