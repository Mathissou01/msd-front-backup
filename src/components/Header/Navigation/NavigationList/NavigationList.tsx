import classNames from "classnames";
import React from "react";
import {
  IServiceLink,
  remapServiceLinksDynamicZone,
} from "../../../../lib/service-links";
import {
  ENavigationPages,
  useNavigation,
} from "../../../../hooks/useNavigation";
import { useContract } from "../../../../hooks/useContract";
import NavigationListButton from "./NavigationListButton/NavigationListButton";
import "./navigation-list.scss";
import { removeNulls } from "../../../../lib/utilities";

interface INavigationListProps {
  isDesktopMode: boolean;
  onNavigationClick?: () => void;
}

export default function NavigationList({
  isDesktopMode,
  onNavigationClick,
}: INavigationListProps) {
  /* Static Data */
  const homepageLabel = "Accueil";
  const defaultPicto = "/images/pictos/default.svg";

  /* Methods */
  function isActiveRoute(route: ENavigationPages | string) {
    return route === currentPage ? "c-NavigationList__Item_active" : "";
  }

  function handleClick(path: string) {
    setCurrentPage(path);
    if (onNavigationClick) {
      onNavigationClick();
    }
  }

  /* External Data */
  const { contract } = useContract();
  const { currentPage, setCurrentPage } = useNavigation();

  const contractMenus: Array<IServiceLink> | null =
    remapServiceLinksDynamicZone(
      contract.attributes?.contractMenu?.data?.attributes?.serviceLinks?.filter(
        removeNulls,
      ) ?? [],
    );

  /* Local Data */
  const dynamicClassNames = classNames("c-NavigationList", {
    "c-NavigationList_desktop": isDesktopMode,
    "c-NavigationList_mobile": !isDesktopMode,
  });

  return (
    <ul className={dynamicClassNames}>
      <li
        className={`c-NavigationList__Item ${isActiveRoute(
          ENavigationPages["HomeLink"],
        )}`}
      >
        <NavigationListButton
          href={ENavigationPages["HomeLink"]}
          label={homepageLabel}
          pictoUrl="/images/pictos/homepage.svg"
          isDesktopMode={isDesktopMode}
          isActive={currentPage === ENavigationPages["HomeLink"]}
          onClick={() => handleClick(ENavigationPages["HomeLink"])}
        />
      </li>
      {/* <li
        className={`c-NavigationList__Item ${isActiveRoute(
          ENavigationPages["MonCompteurDechets"],
        )}`}
      >
        <NavigationListButton
          href={ENavigationPages["MonCompteurDechets"]}
          label="mon compteur dÃ©chets"
          pictoUrl="/images/pictos/mon_compteur_dechets.svg"
          isDesktopMode={isDesktopMode}
          isActive={currentPage === ENavigationPages["MonCompteurDechets"]}
          onClick={() => handleClick(ENavigationPages["MonCompteurDechets"])}
        />
      </li> */}
      {contractMenus &&
        contractMenus.map((menu, index) => {
          if (
            menu?.isDisplayed &&
            menu?.type &&
            menu?.type in ENavigationPages
          ) {
            return (
              <li
                key={`navigation_link_${index}`}
                className={`c-NavigationList__Item ${isActiveRoute(
                  menu.generatedSlug,
                )}`}
              >
                <NavigationListButton
                  href={menu.generatedSlug}
                  label={menu.name}
                  pictoUrl={menu.picto?.data?.attributes?.url ?? defaultPicto}
                  isDesktopMode={isDesktopMode}
                  isActive={currentPage === menu.generatedSlug}
                  onClick={() => handleClick(menu.generatedSlug)}
                />
              </li>
            );
          }
        })}
    </ul>
  );
}
