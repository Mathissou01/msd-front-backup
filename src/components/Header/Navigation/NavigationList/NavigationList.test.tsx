import { render, screen } from "@testing-library/react";
import { ContractEntity } from "../../../../graphql/codegen/generated-types";
import globalMock from "../../../../../__mocks__/globalMock.json";
import { ContractContext } from "../../../../hooks/useContract";
import NavigationList from "./NavigationList";

describe("NavigationList", () => {
  it("renders in mobile mode", async () => {
    const { container } = render(
      <ContractContext.Provider
        value={{
          contract: globalMock.contract as ContractEntity,
          contractId: `${Number.parseInt(globalMock.contract.id)}`,
        }}
      >
        <NavigationList isDesktopMode={false} />
      </ContractContext.Provider>,
    );

    expect(await screen.findByText("Accueil")).toBeInTheDocument();
    expect(await screen.findByText("Guide du tri")).toBeInTheDocument();
    expect(await container).toMatchSnapshot();
  });

  it("renders in desktop mode", async () => {
    const { container } = render(
      <ContractContext.Provider
        value={{
          contract: globalMock.contract as ContractEntity,
          contractId: `${Number.parseInt(globalMock.contract.id)}`,
        }}
      >
        <NavigationList isDesktopMode={true} />
      </ContractContext.Provider>,
    );

    expect(await screen.findByText("Accueil")).toBeInTheDocument();
    expect(await screen.findByText("Guide du tri")).toBeInTheDocument();
    expect(await container).toMatchSnapshot();
  });
});
