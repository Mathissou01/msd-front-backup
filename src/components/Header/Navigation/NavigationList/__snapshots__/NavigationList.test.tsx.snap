// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavigationList renders in desktop mode 1`] = `
<div>
  <ul
    class="c-NavigationList c-NavigationList_desktop"
  >
    <li
      class="c-NavigationList__Item c-NavigationList__Item_active"
    >
      <a
        class="c-NavigationListButton c-NavigationListButton_desktop c-NavigationListButton_active"
        href="/"
      >
        <div
          class="c-NavigationListButton__Circle"
        >
          <div
            class="c-NavigationListButton__Picto"
          >
            <img
              alt=""
              data-nimg="1"
              decoding="async"
              height="24"
              loading="lazy"
              src="/images/pictos/homepage.svg"
              style="color: transparent;"
              width="24"
            />
          </div>
        </div>
        <span
          class="c-NavigationListButton__Label"
        >
          Accueil
        </span>
      </a>
    </li>
    <li
      class="c-NavigationList__Item "
    >
      <a
        class="c-NavigationListButton c-NavigationListButton_desktop"
        href="/service/guide-tri"
      >
        <div
          class="c-NavigationListButton__Circle"
        >
          <div
            class="c-NavigationListButton__Picto"
          >
            <img
              alt=""
              data-nimg="1"
              decoding="async"
              height="24"
              loading="lazy"
              src="/images/pictos/default.svg"
              style="color: transparent;"
              width="24"
            />
          </div>
        </div>
        <span
          class="c-NavigationListButton__Label"
        >
          Guide du tri
        </span>
      </a>
    </li>
    <li
      class="c-NavigationList__Item "
    >
      <a
        class="c-NavigationListButton c-NavigationListButton_desktop"
        href="/actualites"
      >
        <div
          class="c-NavigationListButton__Circle"
        >
          <div
            class="c-NavigationListButton__Picto"
          >
            <img
              alt=""
              data-nimg="1"
              decoding="async"
              height="24"
              loading="lazy"
              src="/images/pictos/default.svg"
              style="color: transparent;"
              width="24"
            />
          </div>
        </div>
        <span
          class="c-NavigationListButton__Label"
        >
          Actualités
        </span>
      </a>
    </li>
    <li
      class="c-NavigationList__Item "
    >
      <a
        class="c-NavigationListButton c-NavigationListButton_desktop"
        href="/evenements"
      >
        <div
          class="c-NavigationListButton__Circle"
        >
          <div
            class="c-NavigationListButton__Picto"
          >
            <img
              alt=""
              data-nimg="1"
              decoding="async"
              height="24"
              loading="lazy"
              src="/images/pictos/default.svg"
              style="color: transparent;"
              width="24"
            />
          </div>
        </div>
        <span
          class="c-NavigationListButton__Label"
        >
          Évenements
        </span>
      </a>
    </li>
    <li
      class="c-NavigationList__Item "
    >
      <a
        class="c-NavigationListButton c-NavigationListButton_desktop"
        href="/contenu-libre/1"
      >
        <div
          class="c-NavigationListButton__Circle"
        >
          <div
            class="c-NavigationListButton__Picto"
          >
            <img
              alt=""
              data-nimg="1"
              decoding="async"
              height="24"
              loading="lazy"
              src="/images/pictos/default.svg"
              style="color: transparent;"
              width="24"
            />
          </div>
        </div>
        <span
          class="c-NavigationListButton__Label"
        >
          Editorial 1
        </span>
      </a>
    </li>
    <li
      class="c-NavigationList__Item "
    >
      <a
        class="c-NavigationListButton c-NavigationListButton_desktop"
        href="/contenu-libre/2"
      >
        <div
          class="c-NavigationListButton__Circle"
        >
          <div
            class="c-NavigationListButton__Picto"
          >
            <img
              alt=""
              data-nimg="1"
              decoding="async"
              height="24"
              loading="lazy"
              src="/images/pictos/default.svg"
              style="color: transparent;"
              width="24"
            />
          </div>
        </div>
        <span
          class="c-NavigationListButton__Label"
        >
          Editorial 2
        </span>
      </a>
    </li>
  </ul>
</div>
`;

exports[`NavigationList renders in mobile mode 1`] = `
<div>
  <ul
    class="c-NavigationList c-NavigationList_mobile"
  >
    <li
      class="c-NavigationList__Item c-NavigationList__Item_active"
    >
      <a
        class="c-NavigationListButton c-NavigationListButton_mobile c-NavigationListButton_active"
        href="/"
      >
        <div
          class="c-NavigationListButton__Circle"
        >
          <div
            class="c-NavigationListButton__Picto"
          >
            <img
              alt=""
              data-nimg="1"
              decoding="async"
              height="24"
              loading="lazy"
              src="/images/pictos/homepage.svg"
              style="color: transparent;"
              width="24"
            />
          </div>
        </div>
        <span
          class="c-NavigationListButton__Label"
        >
          Accueil
        </span>
      </a>
    </li>
    <li
      class="c-NavigationList__Item "
    >
      <a
        class="c-NavigationListButton c-NavigationListButton_mobile"
        href="/service/guide-tri"
      >
        <div
          class="c-NavigationListButton__Circle"
        >
          <div
            class="c-NavigationListButton__Picto"
          >
            <img
              alt=""
              data-nimg="1"
              decoding="async"
              height="24"
              loading="lazy"
              src="/images/pictos/default.svg"
              style="color: transparent;"
              width="24"
            />
          </div>
        </div>
        <span
          class="c-NavigationListButton__Label"
        >
          Guide du tri
        </span>
      </a>
    </li>
    <li
      class="c-NavigationList__Item "
    >
      <a
        class="c-NavigationListButton c-NavigationListButton_mobile"
        href="/actualites"
      >
        <div
          class="c-NavigationListButton__Circle"
        >
          <div
            class="c-NavigationListButton__Picto"
          >
            <img
              alt=""
              data-nimg="1"
              decoding="async"
              height="24"
              loading="lazy"
              src="/images/pictos/default.svg"
              style="color: transparent;"
              width="24"
            />
          </div>
        </div>
        <span
          class="c-NavigationListButton__Label"
        >
          Actualités
        </span>
      </a>
    </li>
    <li
      class="c-NavigationList__Item "
    >
      <a
        class="c-NavigationListButton c-NavigationListButton_mobile"
        href="/evenements"
      >
        <div
          class="c-NavigationListButton__Circle"
        >
          <div
            class="c-NavigationListButton__Picto"
          >
            <img
              alt=""
              data-nimg="1"
              decoding="async"
              height="24"
              loading="lazy"
              src="/images/pictos/default.svg"
              style="color: transparent;"
              width="24"
            />
          </div>
        </div>
        <span
          class="c-NavigationListButton__Label"
        >
          Évenements
        </span>
      </a>
    </li>
    <li
      class="c-NavigationList__Item "
    >
      <a
        class="c-NavigationListButton c-NavigationListButton_mobile"
        href="/contenu-libre/1"
      >
        <div
          class="c-NavigationListButton__Circle"
        >
          <div
            class="c-NavigationListButton__Picto"
          >
            <img
              alt=""
              data-nimg="1"
              decoding="async"
              height="24"
              loading="lazy"
              src="/images/pictos/default.svg"
              style="color: transparent;"
              width="24"
            />
          </div>
        </div>
        <span
          class="c-NavigationListButton__Label"
        >
          Editorial 1
        </span>
      </a>
    </li>
    <li
      class="c-NavigationList__Item "
    >
      <a
        class="c-NavigationListButton c-NavigationListButton_mobile"
        href="/contenu-libre/2"
      >
        <div
          class="c-NavigationListButton__Circle"
        >
          <div
            class="c-NavigationListButton__Picto"
          >
            <img
              alt=""
              data-nimg="1"
              decoding="async"
              height="24"
              loading="lazy"
              src="/images/pictos/default.svg"
              style="color: transparent;"
              width="24"
            />
          </div>
        </div>
        <span
          class="c-NavigationListButton__Label"
        >
          Editorial 2
        </span>
      </a>
    </li>
  </ul>
</div>
`;
